src/server/builder.ts

```diff
@@ -0,0 +1,405 @@
+/// <reference path="..\compiler\commandLineParser.ts" />
+/// <reference path="..\services\services.ts" />
+/// <reference path="protocol.d.ts" />
+/// <reference path="session.ts" />
+/// <reference types="node" />
+
+namespace ts.server {
+
+    const crypto: any = require("crypto");
+
+    /**
+     * An abstract file info that maintains a shape signature.
+     */
+    export class BuilderFileInfo {
+
+        private lastCheckedShapeSignature: string;
+        private lastEmittedContentSignature: string;
+        scriptInfo: ScriptInfo;
+
+        constructor(public fileName: string) {
+        }
+
+        public isExternalModule(builder: Builder) {
+            const sourceFile = this.getSourceFile(builder);
+            if (!!sourceFile.externalModuleIndicator) {
+                return true;
+            }
+            if (sourceFile.isDeclarationFile) {

```

Comment:
strictly speaking this is not correct:
- ambient external modules can be mixed with ambient declarations
- ambient external modules can appear in `.ts` files